File: Data_Structure/Trees/Complete_Binary_Tree.c
#include<stdio.h>
#include<stdbool.h>

struct node
{
    int data;
    struct node* left;
    struct node* right;
};

//Creating a new node
struct node* createNode(int item)
{
    struct node* newNode = (struct node*)malloc(sizeof(struct node));
    newNode -> data = item;
    newNode -> left = NULL;
    newNode -> right = NULL;
    return newNode;
}

//Count the number of nodes 
int countNumNodes(struct node *root) 
{ 
    if(root == NULL) 
    {
        return (0);
    } 
    return(1 + countNumNodes(root->left) + countNumNodes(root->right));
 }

//Check if the tree is complete
bool isCompleteBinaryTree(struct node* root, int index, int countNumNodes)
{
    if(root == NULL)
    {
        return true;
    }
    if(index >= countNumNodes)
    {
        return false;
    }
    return(isCompleteBinaryTree(root -> left, 2*index+1, countNumNodes) && isCompleteBinaryTree(root -> right, 2*index+2, countNumNodes));
}

struct node* insertLeft(struct node* root, int item)
{
    root -> left = createNode(item);
    return root -> left;
}

struct node* insertRight(struct node* root, int item)
{
    root -> right = createNode(item);
    return root -> right;
}

int main()
{
    struct node* root = createNode(2);
    insertLeft(root, 5);
    insertRight(root, 7);
    insertLeft(root -> left, 8);
    insertRight(root -> left, 82);
    insertLeft(root -> right, 10);
    insertRight(root -> right, 29);
    
    int newCount = countNumNodes(root);
    int index = 0;
    
    if(isCompleteBinaryTree(root, index, newCount))
    {
        printf("It is a Complete Binary Tree");
    }
    else
    {
        printf("It is not a Complete Binary Tree");
    }
    
    return 0;
}

File: Data_Structure/DS_Using_Python/Complete_Binary_Tree.py
""" Construct below tree
             1
           /   \
          /     \
         2       3
        / \     / \
       4   5   6   7
    """
class CompleteBinaryTree:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None
		
	#Count the number of nodes 
	def countNumNodes(self, root):
		if root is None:
			return 0
		return (1+self.countNumNodes(root.left) + self.countNumNodes(root.right))
		
	#Check if the tree is complete
	def isCompleteBinaryTree(self, tree, index ,countNumNodes):
		if tree is None:
			return True
		if index >= countNumNodes:
			return False
		return self.isCompleteBinaryTree(tree.left, 2*index+1, countNumNodes) and self.isCompleteBinaryTree(tree.right, 2*index+2, countNumNodes)
		
	#insert data in left child node
	def insertLeft(self, newNode):
		if self.left is None:
			self.left = CompleteBinaryTree(newNode)
		else:
			tree = CompleteBinaryTree(newNode)
			tree.left = self.left
			self.left = tree
		
	#insert data in right child node
	def insertRight(self, newNode):
		if self.right is None:
			self.right = CompleteBinaryTree(newNode)
		else:
			tree = CompleteBinaryTree(newNode)
			tree.right = self.right
			self.right = tree
		
root = CompleteBinaryTree(1)
root.insertLeft(2)
root.insertRight(3)
root.insertLeft(4)
#root.insertRight(5)
#root.insertLeft(6)
#root.insertRight(7)

nodeCount = root.countNumNodes(root)
index = 0

if(root.isCompleteBinaryTree(root, index, nodeCount)):
	print("It is a Complete Binary Tree")
else:
	print("It is not a Complete Binary Tree")